File: frontend/src/modules/math.ts
import { block } from "blockly/core/tooltip";
import Blockly from 'blockly';
import { blockCodeGenerators, generateCodeForBlock } from "../compiler/compile";
import functionTable, { functionCallers } from "../compiler/functionTable";
import { addCategory } from "../toolbox";

addCategory({
    'kind': 'category',
    'name': 'Math',
    'categorystyle': 'math_category',
    'contents': [
        {
            'type': 'math_number',
            'kind': 'block',
            'fields': {
                'NUM': 123,
            },
        },
        {
            'type': 'math_arithmetic',
            'kind': 'block',
            'fields': {
                'OP': 'ADD',
            },
            'inputs': {
                'A': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },
                'B': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_single',
            'kind': 'block',
            'fields': {
                'OP': 'ROOT',
            },
            'inputs': {
                'NUM': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 9,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_trig',
            'kind': 'block',
            'fields': {
                'OP': 'SIN',
            },
            'inputs': {
                'NUM': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 45,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_constant',
            'kind': 'block',
            'fields': {
                'CONSTANT': 'PI',
            },
        },
        {
            'type': 'math_number_property',
            'kind': 'block',
            'fields': {
                'PROPERTY': 'EVEN',
            },
            'inputs': {
                'NUMBER_TO_CHECK': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 0,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_round',
            'kind': 'block',
            'fields': {
                'OP': 'ROUND',
            },
            'inputs': {
                'NUM': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 3.1,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_modulo',
            'kind': 'block',
            'inputs': {
                'DIVIDEND': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 64,
                        },
                    },
                },
                'DIVISOR': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 10,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_constrain',
            'kind': 'block',
            'inputs': {
                'VALUE': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 50,
                        },
                    },
                },
                'LOW': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },
                'HIGH': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 100,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_random_int',
            'kind': 'block',
            'inputs': {
                'FROM': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },
                'TO': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 100,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_random_float',
            'kind': 'block',
        },
        {
            'type': 'math_atan2',
            'kind': 'block',
            'inputs': {
                'X': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },
                'Y': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },
            },
        },
        {
            'type': 'math_map_linear',
            'kind': 'block',
            'inputs': {
                'X1': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 0,
                        },
                    },
                },

                'Y1': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 0,
                        },
                    },
                },

                'X2': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },
                'Y2': {
                    'shadow': {
                        'type': 'math_number',
                        'fields': {
                            'NUM': 1,
                        },
                    },
                },

            },
        },
        {
            'type': 'math_map_temperature',
            'kind': 'block',
        }
    ],
});

blockCodeGenerators.math_arithmetic = (block, buffer, ctx) => {
    const op = block.getFieldValue('OP') as 'ADD' | 'MINUS' | 'MULTIPLY' | 'DIVIDE' | 'POWER';
    const left = generateCodeForBlock('Number', block.getInputTargetBlock('A'), buffer, ctx);
    const right = generateCodeForBlock('Number', block.getInputTargetBlock('B'), buffer, ctx);
    return { type: 'Number', code: buffer.startSegment(code => functionCallers.mathBinary(code, left, right, op)) }
}

blockCodeGenerators.math_number = (block, buffer, ctx) => {
    return { type: 'Number', code: buffer.startSegment().addPushFloat(parseFloat(block.getFieldValue('NUM'))) }
}

blockCodeGenerators.math_modulo = (block, buffer, ctx) => {
    const dividend = generateCodeForBlock('Number', block.getInputTargetBlock('DIVIDEND'), buffer, ctx);
    const divisor = generateCodeForBlock('Number', block.getInputTargetBlock('DIVISOR'), buffer, ctx);
    return { type: 'Number', code: buffer.startSegment(code => functionCallers.mathBinary(code, dividend, divisor, 'MODULO')) }
}

blockCodeGenerators.math_constant = (block, buffer, ctx) => {
    const code = buffer.startSegment();
    switch (block.getFieldValue('CONSTANT')) {
        case 'PI':
            code.addPushFloat(Math.PI);
            break;
        case 'E':
            code.addPushFloat(Math.E);
            break;
        case 'GOLDEN_RATIO':
            code.addPushFloat(1.61803398875);
            break;
        case 'SQRT2':
            code.addPushFloat(Math.SQRT2);
            break;
        case 'SQRT1_2':
            code.addPushFloat(Math.SQRT1_2);
            break;
        case 'INFINITY':
            code.addPushFloat(Infinity);
            break;

    }
    return { type: 'Number', code }
}

blockCodeGenerators.math_number_property = (block, buffer, ctx) => {
    const number = generateCodeForBlock('Number', block.getInputTargetBlock('NUMBER_TO_CHECK'), buffer, ctx);
    const property = block.getFieldValue('PROPERTY') as 'EVEN' | 'ODD' | 'PRIME' | 'WHOLE' | 'POSITIVE' | 'NEGATIVE' | 'DIVISIBLE_BY';
    const code = buffer.startSegment();
    if (property === 'DIVISIBLE_BY') {
        const divisor = generateCodeForBlock('Number', block.getInputTargetBlock('DIVISOR'), buffer, ctx);
        const remainder = { type: 'Number', code: buffer.startSegment(code => functionCallers.mathBinary(code, number, divisor, "MODULO")) } as const
        functionCallers.logicCompare(code, remainder, { type: 'Number', value: 0 }, 'EQ');
    }
    else {
        code.addCall(functionTable.mathNumberProperty, 'Boolean', number, { type: 'uint8', value: { EVEN: 0, ODD: 1, PRIME: 2, WHOLE: 3, POSITIVE: 4, NEGATIVE: 5 }[property] })
    }

    return { type: 'Boolean', code }
}

blockCodeGenerators.math_trig = (block, buffer, ctx) => {
    const op = block.getFieldValue('OP') as 'SIN' | 'COS' | 'TAN' | 'ASIN' | 'ACOS' | 'ATAN';
    const num = generateCodeForBlock('Number', block.getInputTargetBlock('NUM'), buffer, ctx);
    return { type: 'Number', code: buffer.startSegment(code => functionCallers.mathUnary(code, num, op)) }
}

blockCodeGenerators.math_round = (block, buffer, ctx) => {
    const op = block.getFieldValue('OP') as 'ROUND' | 'ROUNDUP' | 'ROUNDDOWN';
    const num = generateCodeForBlock('Number', block.getInputTargetBlock('NUM'), buffer, ctx);
    return { type: 'Number', code: buffer.startSegment(code => functionCallers.mathUnary(code, num, op)) }
}

blockCodeGenerators.math_single = (block, buffer, ctx) => {
    const op = block.getFieldValue('OP') as 'ROOT' | 'ABS' | 'NEG' | 'LN' | 'LOG10' | 'EXP' | 'POW10';
    const num = generateCodeForBlock('Number', block.getInputTargetBlock('NUM'), buffer, ctx);
    buffer.startSegment();
    return { type: 'Number', code: buffer.startSegment(code => functionCallers.mathUnary(code, num, op)) }
}

blockCodeGenerators.math_random_float = (block, buffer, ctx) => {
    return { type: 'Number', code: buffer.startSegment().addCall(functionTable.mathRandomFloat, 'Number') }
}

blockCodeGenerators.math_random_int = (block, buffer, ctx) => {
    const left = generateCodeForBlock('Number', block.getInputTargetBlock('FROM'), buffer, ctx);
    const right = generateCodeForBlock('Number', block.getInputTargetBlock('TO'), buffer, ctx);
    return { type: 'Number', code: buffer.startSegment(code => functionCallers.mathBinary(code, left, right, "RANDOM_INT")) }
}

blockCodeGenerators.math_atan2 = (block, buffer, ctx) => {
    const left = generateCodeForBlock('Number', block.getInputTargetBlock('X'), buffer, ctx);
    const right = generateCodeForBlock('Number', block.getInputTargetBlock('Y'), buffer, ctx);
    return { type: 'Number', code: buffer.startSegment(code => functionCallers.mathBinary(code, left, right, "ATAN2")) }
}

blockCodeGenerators.math_constrain = (block, buffer, ctx) => {
    const value = generateCodeForBlock('Number', block.getInputTargetBlock('VALUE'), buffer, ctx);
    const low = generateCodeForBlock('Number', block.getInputTargetBlock('LOW'), buffer, ctx);
    const high = generateCodeForBlock('Number', block.getInputTargetBlock('HIGH'), buffer, ctx);
    return { type: 'Number', code: buffer.startSegment().addCall(functionTable.mathConstrain, 'Number', value, low, high) }
}

Blockly.Blocks['math_map_linear'] = {
    init: function () {
        this.appendValueInput("VALUE")
            .setCheck("Number")
            .appendField("Map Linear");
        this.appendEndRowInput();
        this.appendValueInput("X1")
            .setCheck("Number")
            .appendField("In1");
        this.appendValueInput("Y1")
            .setCheck("Number")
            .appendField("Out1");
        this.appendEndRowInput();
        this.appendValueInput("X2")
            .setCheck("Number")
            .appendField("In2");
        this.appendValueInput("Y2")
            .setCheck("Number")
            .appendField("Out2");
        this.setOutput(true, "Number");

        this.setColour(230);
        this.setTooltip("");
        this.setHelpUrl("");
    }
};

blockCodeGenerators.math_map_linear = (block, buffer, ctx) => {
    return {
        type: 'Number', code: buffer.startSegment().addCall(functionTable.mathMapLinear, 'Number',
            generateCodeForBlock('Number', block.getInputTargetBlock('VALUE'), buffer, ctx),
            generateCodeForBlock('Number', block.getInputTargetBlock('X1'), buffer, ctx),
            generateCodeForBlock('Number', block.getInputTargetBlock('Y1'), buffer, ctx),
            generateCodeForBlock('Number', block.getInputTargetBlock('X2'), buffer, ctx),
            generateCodeForBlock('Number', block.getInputTargetBlock('Y2'), buffer, ctx),
        )
    };
};

Blockly.Blocks['math_map_temperature'] = {
    init: function () {
        this.appendValueInput("VALUE")
            .setCheck("Number")
            .appendField("Map Temperature");
        this.appendEndRowInput();
        this.appendEndRowInput()
            .appendField("Value 1")
            .appendField(new Blockly.FieldNumber(0.4), "X1")
            .appendField("Temp 1")
            .appendField(new Blockly.FieldNumber(20), "T1");
        this.appendEndRowInput()
            .appendField("Value 2")
            .appendField(new Blockly.FieldNumber(0.5), "X2")
            .appendField("Temp 2")
            .appendField(new Blockly.FieldNumber(30), "T2");
        this.setOutput(true, "Number");
        this.setColour(230);
        this.setTooltip("");
        this.setHelpUrl("");
    }
};

function toResistance(adcFraction: number) {
    if (Math.abs(1 - adcFraction) < 1e-6) {
        return 1e5;
    }
    return (adcFraction) / (1 - adcFraction);
}

blockCodeGenerators.math_map_temperature = (block, buffer, ctx) => {

    // steinhart-hart equation
    const logR1 = Math.log(toResistance(block.getFieldValue('X1')));
    const logR2 = Math.log(toResistance(block.getFieldValue('X2')));
    const t1_inv = 1 / (block.getFieldValue('T1') + 273.15);
    const t2_inv = 1 / (block.getFieldValue('T2') + 273.15);
    const logDiff = logR2 - logR1;

    let a: number
    let b: number
    if (Math.abs(logDiff) < 1e-6) {
        a = 1;
        b = 1;
    }
    else {
        b = (t2_inv - t1_inv) / (logR2 - logR1);
        a = t1_inv - logR1 * b;
    }

    return {
        type: 'Number', code: buffer.startSegment().addCall(functionTable.mathMapTemperature, 'Number',
            generateCodeForBlock('Number', block.getInputTargetBlock('VALUE'), buffer, ctx),
            { type: 'Number', value: a },
            { type: 'Number', value: b },
        )
    };
};

File: esp32/src/micro-blocks/modules/math.cpp
#include "logic.h"
#include "../machine.h"
#include <Arduino.h>

namespace mathModule
{

    bool isPrime(long number)
    {
        if (number < 2)
            return false;
        if (number == 2)
            return true;
        if (number % 2 == 0)
            return false;
        for (long i = 3; (i * i) <= number; i += 2)
        {
            if (number % i == 0)
                return false;
        }
        return true;
    }

    float toResistance(float adcFraction)
    {
        // calculate resistance of thermistor
        // r1=thermistor (ADC-GND)
        // r2=reference (AREF/ADC)
        // i1=i2; u=r*i; i=u/r; u1/r1=u2/r2; r1/u1=r2/u2; r1=r2*u1/u2
        // u2=AREF-u1;
        // u1=AREF*ADC
        // r1=r2*AREF*ADC/(AREF-AREF*ADC)
        // r1=r2*ADC/(1-ADC)
        if (abs(1 - adcFraction) < 1e-6)
        {
            return 1e5;
        }
        return (adcFraction) / (1 - adcFraction);
    }

    void setup()
    {
        // mathBinary
        machine::registerFunction(
            6,
            []()
            {
                auto operation = machine::popUint8();
                auto right = machine::popFloat();
                auto left = machine::popFloat();

                float result;
                switch (operation)
                {
                case 0:
                    result = left + right;
                    break;
                case 1:
                    result = left - right;
                    break;
                case 2:
                    result = left * right;
                    break;
                case 3:
                    result = left / right;
                    break;
                case 4:
                    result = pow(left, right);
                    break;
                case 5:
                    result = fmod(left, right);
                    break;
                case 6:
                    result = random(left, right);
                    break;

                case 7:
                    result = atan2(left, right);
                    break;

                default:
                    Serial.println("Invalid operation for mathArithmetic");
                }

                // Serial.println(String("mathArithmetic: ") + left + " [" + operation + "] " + right + " = " + result);
                machine::pushFloat(result);
            });

        // mathNumberProperty
        machine::registerFunction(
            15,
            []()
            {
                auto property = machine::popUint8();
                auto number = machine::popFloat();
                bool result;
                switch (property)
                {
                case 0:
                    result = ((int)number % 2) == 0;
                    break;
                case 1:
                    result = ((int)number % 2) == 1;
                    break;
                case 2:
                    result = isPrime((long)number);
                    break;
                case 3:
                    result = ((int)number % 1) == 0;
                    break;
                case 4:
                    result = number > 0;
                    break;
                case 5:
                    result = number < 0;
                    break;
                default:
                    Serial.println("Invalid property for mathNumberProperty");
                }

                Serial.println(String("mathNumberProperty: ") + number + " [" + property + "] = " + result);
                machine::pushUint8(result);
            });

        // unary operations
        machine::registerFunction(
            16,
            []()
            {
                auto operation = machine::popUint8();
                auto number = machine::popFloat();
                float result;
                switch (operation)
                {
                // trig
                case 0:
                    result = sin(number);
                    break;
                case 1:
                    result = cos(number);
                    break;
                case 2:
                    result = tan(number);
                    break;
                case 3:
                    result = asin(number);
                    break;
                case 4:
                    result = acos(number);
                    break;
                case 5:
                    result = atan(number);
                    break;

                // round
                case 6:
                    result = round(number);
                    break;
                case 7:
                    result = ceil(number);
                    break;
                case 8:
                    result = floor(number);
                    break;

                // single
                case 9:
                    result = sqrt(number);
                    break;
                case 10:
                    result = abs(number);
                    break;
                case 11:
                    result = -number;
                    break;
                case 12:
                    result = log(number);
                    break;
                case 13:
                    result = log10(number);
                    break;
                case 14:
                    result = exp(number);
                    break;
                case 15:
                    result = pow(10, number);
                    break;
                default:
                    Serial.println("Invalid operation for mathTrig");
                }

                // Serial.println(String("mathTrig: ") + operation + " " + number + " = " + result);
                machine::pushFloat(result);
            });

        // mathRandomFloat
        machine::registerFunction(
            17,
            []()
            {
                machine::pushFloat(rand() / (float)RAND_MAX);
            });

        // mathConstrain
        machine::registerFunction(
            18,
            []()
            {
                auto high = machine::popFloat();
                auto low = machine::popFloat();
                auto number = machine::popFloat();
                if (number < low)
                    number = low;
                if (number > high)
                    number = high;
                machine::pushFloat(number);
            });

        // mathMapLinear
        machine::registerFunction(
            34,
            []()
            {
                auto y2 = machine::popFloat();
                auto x2 = machine::popFloat();
                auto y1 = machine::popFloat();
                auto x1 = machine::popFloat();
                auto value = machine::popFloat();

                auto xDiff = x2 - x1;
                if (xDiff == 0)
                {
                    machine::pushFloat(value);
                    return;
                }
                machine::pushFloat(y1 + (value - x1) * (y2 - y1) / xDiff);
            });

        // mathMapTemperature
        machine::registerFunction(
            35,
            []()
            {
                auto b = machine::popFloat();
                auto a = machine::popFloat();
                auto value = machine::popFloat();

                machine::pushFloat(1. / (a + b * log(toResistance(value))) - 273.15);
            });
    }

}